cardano-cli/src/Cardano/CLI/Run/Debug/CheckNodeConfiguration.hs

```diff
@@ -0,0 +1,188 @@
+  nodeConfigObj' <- foldM (runAnalysis configFile fixTheFile) nodeConfigObj analyses
+  when (nodeConfigObj' /= nodeConfigObj) $ do
+    -- Write the fixed node configuration file
+    -- Better to remove the file first, to avoid a file busy exception when writing
+    liftIO $ removeFile configFilePath
+    -- Cheap trick to try to preserve the input file's format
+    let looksLikeJson = ".json" `isSuffixOf` configFilePath
+    liftIO $
+      if looksLikeJson
+        then LBS.writeFile configFilePath $ Aeson.encodePretty nodeConfigObj'
+        else Yaml.encodeFile configFilePath nodeConfigObj'
+  liftIO $ putStrLn $ "Successfully checked node configuration file: " <> configFilePath
+ where
+  configFilePath = unFile configFile
+  eras :: [AnyCardanoEra] = [minBound .. maxBound]
+
+data Analysis = Analysis
+  { concernedEra :: AnyCardanoEra
+  , genesisFilepath :: FilePath
+  -- ^ The path to the genesis file, as found in the node configuration file, if any
+  , actualHash :: Text.Text
+  -- ^ The actual hash of the genesis file, computed from the file's content
+  , expectedHash :: Maybe Text.Text
+  -- ^ The hash of the genesis file, as found in the node configuration file, if any.
+  }
+
+runAnalysis
+  :: NodeConfigFile 'In
+  -- ^ The node configuration file. It's not read by this function, but used for producing error messages.
+  -> FixNodeConfigFile
+  -> Aeson.Object
+  -- ^ The node configuration file, as a JSON object. It will get written back to disk.
+  -> Analysis
+  -> ExceptT DebugCmdError IO Aeson.Object
+  -- ^ The fixed node configuration file
+runAnalysis configFile fixMode nodeConfigObj (Analysis{concernedEra = era@(AnyCardanoEra cEra), genesisFilepath = _, actualHash, expectedHash}) = do
+  case (expectedHash, fixMode) of
+    (Nothing, _) ->
+      -- Genesis hash is not declared in the node configuration file: we have nothing to check against
+      pure nodeConfigObj
+    (Just expectedHash', _)
+      | expectedHash' == actualHash ->
+          -- Genesis hash is correct: no error and nothing to change
+          pure nodeConfigObj
+    (Just expectedHash', DontFix) ->
+      -- Genesis hash is incorrect and we don't want to fix it, so we report an error
+      throwError $ DebugNodeConfigWrongGenesisHashCmdError configFilePath era actualHash expectedHash'
+    (Just expectedHash', FixInPlace) -> do
+      liftIO $
+        Text.putStrLn $
+          "Fixing " <> eraToStringKey cEra <> "'s genesis hash in " <> Text.pack configFilePath
+      liftIO $ Text.putStrLn $ "Replacing " <> expectedHash' <> " with " <> actualHash
+      pure $
+        Aeson.insert (Aeson.fromText $ eraToGenesisHashKey era) (Aeson.String actualHash) nodeConfigObj
+ where
+  configFilePath = unFile configFile
+
+-- | Build an 'Analysis' value from the given node configuration file, for the given era.
+-- Reads the genesis file path and hash from the node configuration file, if they exist.
+mkAnalysis
+  :: AnyCardanoEra
```

Comment:
Use the experimental api interface here. We don't care about previous eras.

Reply:
This parameter isn't about the era we support, it's about the era whose hash we want to validate. I updated the haddock to make this explicit, as this indeed is an usual use case for us.

Reply:
We only care about mainnet and the upcoming era and therefore the configuration files associated with those eras which include Byron, Shelley, Alonzo and Conway hashes. I.e we check all of those hashes in the Conway era. If another genesis file is added in an upcoming era then we would add that additional check for the upcoming era. Currently there are a bunch of irrelevant cases that return `Nothing` which we can totally exclude by not parameterizing on `AnyCardanoEra`.